q2_cutadapt/_demux.py

```diff
@@ -122,39 +122,80 @@ def _write_empty_fastq_to_mux_barcode_in_seq_fmt(seqs_dir_fmt):
         seqs_dir_fmt.file.write_data(fastq, FastqGzFormat)
 
 
-def _demux(seqs, per_sample_sequences, forward_barcodes, reverse_barcodes,
-           error_tolerance, mux_fmt, batch_size, minimum_length, cores):
-    fwd_barcode_name = forward_barcodes.name
-    forward_barcodes = forward_barcodes.drop_missing_values()
-    barcodes = forward_barcodes.to_series().to_frame()
+def _check_barcodes_uniqueness(
+        forward_barcodes: qiime2.CategoricalMetadataColumn,
+        reverse_barcodes: qiime2.CategoricalMetadataColumn = None,
+        mixed_orientation: bool = False,
+):
+    for_barcodes = forward_barcodes.to_series().to_frame()
     if reverse_barcodes is not None:
-        barcode_pairs = set()
-        samples_w_missing_barcodes = set()
-        samples_w_dup_barcode_pairs = set()
-        rev_barcode_name = reverse_barcodes.name
         rev_barcodes = reverse_barcodes.to_series()
-        # 'sort = false' below prevents a warning about future behavior changes
-        # by selecting the future behavior explicitly
-        barcodes = pd.concat([barcodes, rev_barcodes], axis=1, sort=False)
-
-        for sample_id, f_barcode, r_barcode in barcodes.itertuples():
+    else:
+        # Quite ugly but it allows to continue to use the same logic as before
+        rev_barcodes = forward_barcodes.to_series()
```

Comment:
As the `rev_barcodes` is not used if `reverse_barcodes` was not provided, let's say it should not be unsafe.
Alternatively, I could create a `Series` of the same length than `forward_barcodes` and filled with `None`. By doing so, I could also later use `if r_barcode is None:` rather than the current `if reverse_barcodes is None:` which could make the code clearer

Reply:
In the end, I ended up branching at the very first `if reverse_barcodes is not None:` and separate both logics, which might be clearer for future readers